File: python/grpcproxy/backend.py
import asyncio

import grpc

from grpcproxy.generated import echo_pb2, echo_pb2_grpc, math_pb2, math_pb2_grpc


class EchoServiceServicer(echo_pb2_grpc.EchoServiceServicer):
    async def UnaryEcho(self, request, context):
        return echo_pb2.EchoResponse(message=request.message, source="backend")

    async def ServerStreamEcho(self, request, context):
        for i in range(3):
            yield echo_pb2.EchoResponse(
                message=f"{request.message} #{i + 1}", source="backend"
            )

    async def ClientStreamEcho(self, request_iterator, context):
        messages = []
        async for req in request_iterator:
            messages.append(req.message)
        return echo_pb2.EchoResponse(message=" ".join(messages), source="backend")

    async def BidiStreamEcho(self, request_iterator, context):
        async for req in request_iterator:
            yield echo_pb2.EchoResponse(message=req.message, source="backend")


class MathServiceServicer(math_pb2_grpc.MathServiceServicer):
    async def Add(self, request, context):
        if request.a == 0 and request.b == 0:
            await context.abort(grpc.StatusCode.INVALID_ARGUMENT, "both operands are zero")
        return math_pb2.AddResponse(result=request.a + request.b, source="backend")

    async def Fibonacci(self, request, context):
        a, b = 0, 1
        for _ in range(request.count):
            yield math_pb2.FibResponse(value=a, source="backend")
            a, b = b, a + b


async def start_backend(addr: str) -> tuple[grpc.aio.Server, int]:
    server = grpc.aio.server()
    echo_pb2_grpc.add_EchoServiceServicer_to_server(EchoServiceServicer(), server)
    math_pb2_grpc.add_MathServiceServicer_to_server(MathServiceServicer(), server)
    port = server.add_insecure_port(addr)
    await server.start()
    return server, port

File: backend/backend.go
package backend

import (
	"context"
	"fmt"
	"io"
	"net"
	"strings"

	"github.com/tordynnar/grpcproxy/pb"
	"google.golang.org/grpc"
)

type echoServer struct {
	pb.UnimplementedEchoServiceServer
}

func (s *echoServer) UnaryEcho(_ context.Context, req *pb.EchoRequest) (*pb.EchoResponse, error) {
	return &pb.EchoResponse{Message: req.Message, Source: "backend"}, nil
}

func (s *echoServer) ServerStreamEcho(req *pb.EchoRequest, stream grpc.ServerStreamingServer[pb.EchoResponse]) error {
	for i := 0; i < 3; i++ {
		if err := stream.Send(&pb.EchoResponse{
			Message: fmt.Sprintf("%s #%d", req.Message, i+1),
			Source:  "backend",
		}); err != nil {
			return err
		}
	}
	return nil
}

func (s *echoServer) ClientStreamEcho(stream grpc.ClientStreamingServer[pb.EchoRequest, pb.EchoResponse]) error {
	var messages []string
	for {
		req, err := stream.Recv()
		if err == io.EOF {
			return stream.SendAndClose(&pb.EchoResponse{
				Message: strings.Join(messages, " "),
				Source:  "backend",
			})
		}
		if err != nil {
			return err
		}
		messages = append(messages, req.Message)
	}
}

func (s *echoServer) BidiStreamEcho(stream grpc.BidiStreamingServer[pb.EchoRequest, pb.EchoResponse]) error {
	for {
		req, err := stream.Recv()
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		if err := stream.Send(&pb.EchoResponse{
			Message: req.Message,
			Source:  "backend",
		}); err != nil {
			return err
		}
	}
}

type mathServer struct {
	pb.UnimplementedMathServiceServer
}

func (s *mathServer) Add(_ context.Context, req *pb.AddRequest) (*pb.AddResponse, error) {
	return &pb.AddResponse{Result: req.A + req.B, Source: "backend"}, nil
}

func (s *mathServer) Fibonacci(req *pb.FibRequest, stream grpc.ServerStreamingServer[pb.FibResponse]) error {
	a, b := int64(0), int64(1)
	for i := int32(0); i < req.Count; i++ {
		if err := stream.Send(&pb.FibResponse{Value: a, Source: "backend"}); err != nil {
			return err
		}
		a, b = b, a+b
	}
	return nil
}

func StartServer(addr string) (*grpc.Server, net.Listener, error) {
	lis, err := net.Listen("tcp", addr)
	if err != nil {
		return nil, nil, fmt.Errorf("listen: %w", err)
	}
	srv := grpc.NewServer()
	pb.RegisterEchoServiceServer(srv, &echoServer{})
	pb.RegisterMathServiceServer(srv, &mathServer{})
	go srv.Serve(lis)
	return srv, lis, nil
}
